### PR TITLE
beets: new port

### PR DESCRIPTION
**Pull request by catap**
#### Description

This is port for [beets](https://beets.io) a nice library to manage music collections.

Thus, it also includes a lot of subports like `beets-*` to install some 3rd party plugins. Hinestly, I've added all [other-plugins](https://beets.readthedocs.io/en/stable/plugins/#other-plugins) which is mentored in the documentation.

I haven't split it to another 20+ ports because all of them needs a hack to cleanup `beetsplug`, all of them depends on beets and python version.

Anyway, this port depends on:
 - https://github.com/macports/macports-ports/pull/12112
 - https://github.com/macports/macports-ports/pull/12113
 - https://github.com/macports/macports-ports/pull/12114
 - https://github.com/macports/macports-ports/pull/12115
 - https://github.com/macports/macports-ports/pull/12116
 - https://github.com/macports/macports-ports/pull/12117
 - https://github.com/macports/macports-ports/pull/12118
 - https://github.com/macports/macports-ports/pull/12119
 - https://github.com/macports/macports-ports/pull/12120
 - https://github.com/macports/macports-ports/pull/12124
 - https://github.com/macports/macports-ports/pull/12125
 - https://github.com/macports/macports-ports/pull/12126
 - https://github.com/macports/macports-ports/pull/12127
 - https://github.com/macports/macports-ports/pull/12128
 - https://github.com/macports/macports-ports/pull/12129
 - https://github.com/macports/macports-ports/pull/12130
 - https://github.com/macports/macports-ports/pull/12131
 - #12137
 - #12138
 - #12139
 - #12140
 - #12141
 - #12142
 - #12143 
 - #12144
 - #12146
 - #12149
 - #12150
 - #12197
 - #12199
 - #12289

###### Type(s)

- [ ] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on
macOS 11.5.2 20G95 x86_64
Xcode 12.5.1 12E507

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? 
- [x] checked your Portfile with `port lint`?
- [ ] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

### CODE REVIEW

**Comment by pmetzger** [comment]
Quite a lot of subports. Would it make sense just to include all the plugins by default instead, or do they change the behavior enough that a user would want to pick and choose?

**Comment by catap** [comment]
@pmetzger the tricky part that all this plugins came from different pypi or GitHub artifacts. Make everything as single port is fine, but I have no idea how to fetch multiple artifacts.

Thus, different plugins has different versions and may be released at its own time

**Comment by pmetzger** [comment]
BTW, someone familiar with the python portgroup should probably review this.

File: audio/beets/Portfile [hunk 1]
```diff
@@ -0,0 +1,518 @@
...
+                    port:py${python.version}-last \
+                    port:py${python.version}-langdetect \
+                    port:py${python.version}-mpd2 \
+                    port:py${python.version}-flask \
+                    port:py${python.version}-flask-cors \
+                    port:py${python.version}-rarfile \
+                    port:py${python.version}-xdg \
+                    port:py${python.version}-soco \
+                    port:py${python.version}-mutagen \
+                    port:py${python.version}-reflink
```
Comment (reneeotten) [inline audio/beets/Portfile:60]: please sort the dependencies alphabetically

File: audio/beets/Portfile [hunk 2]
```diff
@@ -0,0 +1,518 @@
...
+    post-destroot {
+        xinstall -d ${destroot}${prefix}/share/man/man1
+        xinstall -m 0644 {*}[glob ${worksrcpath}/man/*.1] ${destroot}${prefix}/share/man/man1/
+
+        xinstall -d ${destroot}${prefix}/share/man/man5
+        xinstall -m 0644 {*}[glob ${worksrcpath}/man/*.5] ${destroot}${prefix}/share/man/man5/
+
+        xinstall -d ${destroot}${prefix}/share/${name}
+        xinstall -m 0644 ${worksrcpath}/README.rst ${destroot}${prefix}/share/${name}
+        xinstall -m 0644 ${worksrcpath}/LICENSE ${destroot}${prefix}/share/${name}
```
Comment (reneeotten) [inline audio/beets/Portfile:79]: you can use `xinstall -m 0644 -W ${worksrcpath} file1 file2 ${destroot}${docdir}`

File: audio/beets/Portfile [hunk 3]
```diff
@@ -0,0 +1,518 @@
...
+supported_archs     noarch
+license             MIT
+
+description         music library manager and MusicBrainz tagger
+long_description    Beets is the media library management system for obsessive music geeks. \
+                    The purpose of beets is to get your music collection right once and for all. \
+                    It catalogs your collection, automatically improving its metadata as it goes. \
+                    It then provides a bouquet of tools for manipulating and accessing your music.
+
+homepage            http://beets.io/
```
Comment (reneeotten) [inline audio/beets/Portfile:21]: please use `https` here and elsewhere in the Portfile

**Comment by reneeotten** [comment]
> @pmetzger the tricky part that all this plugins came from different pypi or GitHub artifacts. Make everything as single port is fine, but I have no idea how to fetch multiple artifacts.

I agree that there are a lot of subports, but think the alternative would be to have them all in separate Portfiles which doesn't help all that much. In a way it's nice to have this all confined within one Portfile.

> Thus, different plugins has different versions and may be released at its own time.

To be honest I am not really sure whether it's possible with the current python PG to fetch/build/install multiple distfiles at the same time; one would need to check. 

> And all 3rd party plugins shares the same pice of code:
> 
> ```
>     post-destroot {
>         foreach f {__init__.py __init__.pyc __init__.pyo __pycache__} {
>             set fp "${destroot}${python.pkgd}/beetsplug/$f"
>             if {[file exists ${fp}]} {
>                 file delete -force ${fp}
>             }
>         }
>     }
> ```
> 
> and I have no idea hot to not copy and paste it.

The way that was solved for the `py-backports-<name>` files which also all install in the same directory, was to introduce a port called `py-backports` and have all the `py-backports-<name>` have that listed as `depends_run`. In other words, you do need to create the namespace directory at some point and then individual ports don't have to do that anymore (if they did, it would be removed when uninstalling the port). 

So I guess in this case it would be an option to just create the directory and `__init__.py` file (and whatever else is needed) in the main `beets` port. If you keep the current `post-destroot` where you remove these files for the  individual plugins I think it would work.

**Comment by catap** [comment]
@reneeotten the issue with `__init__.py` that each plugin creates such files and all this files is the same. All ports need `post-destroot` which is removed this file to prevent conflict.

And it works, but as soon as I split it to multiple `Portfile` the nightmare is here.

**Comment by reneeotten** [comment]
> @reneeotten the issue with `__init__.py` that each plugin creates such files and all this files is the same. All ports need `post-destroot` which is removed this file to prevent conflict.

Yes, I understand that and that's what happens now in the way you have defined the `post-destroot` phase that will be run for all plugins.
 
> And it works, but as soon as I split it to multiple `Portfile` the nightmare is here.

Not necessarily, you can check how this was solved for the `py-backports-<>` family of packages as I pointed out earlier. However, I am not suggestion that you should separate out all the plugins into individual ports - I am personally fine with how you have done it. 

The only "issue" I am possibly seeing is that you now *always* remove the `__init__.py` file, which isn't correct either I think. In other words, none of the plugins installs the `__init__.py` file in the `beetsplug` directory in the way you have implemented it now - correct? Or is this already generated by the main `beets` port - can you please check that? But again, I haven't build the port locally to verify that this is indeed the case, might have some time in the weekend to do that.

**Comment by catap** [comment]
@reneeotten thanks for your time in advance!

I'd like confirm that `beets` port install required file which I removed:
```
√ macports-ports % port contents beets | grep 'beetsplug/__init'
  /opt/local/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/beetsplug/__init__.py
√ macports-ports %  
```

All plugins contains such file with content like:
```
from pkgutil import extend_path
__path__ = extend_path(__path__, __name__)
```
with the only goal to make this a namespace package.

File: audio/beets/Portfile [hunk 4]
```diff
@@ -0,0 +1,517 @@
...
+}
+
+subport beets-usertag {
+    PortGroup       github 1.0
+
+    github.setup    igordertigor beets-usertag 8e058361c9dfa335980048c0afb9274e5f077611
+    version         20210908
+    revision        0
+
+    license         Unknown
```
Comment (reneeotten) [inline audio/beets/Portfile:426]: according to upstream this should be `MIT`:
https://github.com/igordertigor/beets-usertag/blob/8e058361c9dfa335980048c0afb9274e5f077611/setup.py#L11`

can you please make sure

File: audio/beets/Portfile [hunk 5]
```diff
@@ -0,0 +1,517 @@
...
+    depends_lib-append \
+                    port:py${python.version}-alive-progress
+}
+
+subport beets-ibroadcast {
+    name            beets-ibroadcast
+    version         0.5.0
+    revision        0
+
+    license         Unlicense
```
Comment (reneeotten) [inline audio/beets/Portfile:300]: ideally all licenses should be recognized by MacPorts (see https://github.com/macports/macports-infrastructure/blob/master/jobs/distributable_lib.tcl); I don't think this one is... But I *think* that `public-domain` would be an appropriate choice here

File: audio/beets/Portfile [hunk 6]
```diff
@@ -0,0 +1,517 @@
...
+}
+
+subport beets-originquery {
+    PortGroup       github 1.0
+
+    github.setup    x1ppy beets-originquery 269e9e4cfc88128493b25ed814e34df2528dc1ef
+    version         20210908
+    revision        0
+
+    license         Unknown
```
Comment (reneeotten) [inline audio/beets/Portfile:382]: It's not clear indeed what's the intended license - you could ask upstream about it

File: audio/beets/Portfile (hunk 4 above)
Comment (catap) [reply audio/beets/Portfile:426]: @reneeotten you're right, I haven't checked `setup.py`.

**Comment by catap** [comment]
@reneeotten sorry, I've switched all my workflow to this port and discovered that I forgot a few plugins.

Just pushed all of them.

File: audio/beets/Portfile [hunk 7]
```diff
@@ -0,0 +1,601 @@
...
+            set fp "${destroot}${python.pkgd}/beetsplug/$f"
+            if {[file exists ${fp}]} {
+                file delete -force ${fp}
+            }
+        }
+    }
+}
+
+subport beets-alternatives {
+    name            beets-alternatives
```
Comment (reneeotten) [inline audio/beets/Portfile:95]: for every subport you get a `port lint` message like: `Error: Portfile directory beets does not match port name beets-alternatives`. 

You should remove the `name` field in every subport as that is derived from the `subport <name>` already and if you do, it is fine with being in the Portfile in the `beets` directory.

File: audio/beets/Portfile [hunk 8]
```diff
@@ -0,0 +1,601 @@
...
+    version         0.10.2.post1
+    revision        0
+
+    license         MIT
+
+    description     beets plugin to manage multiple files
+    long_description ${description}
```
Comment (reneeotten) [inline audio/beets/Portfile:102]: the indentation for `long_description` isn't right in every subport; it doesn't follow the "multiple of four spaces"

File: audio/beets/Portfile [hunk 9]
```diff
@@ -0,0 +1,601 @@
...
+    depends_lib-append \
+                    port:py${python.version}-bottlenose
+
+    build {}
+
+    destroot {
+        xinstall -d ${destroot}${prefix}/share/${name}
+        xinstall -m 0644 ${worksrcpath}/README.md ${destroot}${prefix}/share/${name}
+
+        xinstall -d ${destroot}${python.pkgd}/beetsplug
```
Comment (reneeotten) [inline audio/beets/Portfile:138]: I thought this directory was already made by the `beets` main port, correct? If so, there is no need to do this.

File: audio/beets/Portfile (hunk 5 above)
Comment (reneeotten) [reply audio/beets/Portfile:350]: you didn't agree with the suggestion to use `public-domain`?

File: audio/beets/Portfile [hunk 10]
```diff
@@ -0,0 +1,601 @@
...
+
+    github.setup    edgars-supe beets-importreplace 0.1.1 v
+    revision        0
+
+    license         MIT
+
+    description     Plugin for beets to perform regex replacements during import
+    long_description ${description}
+
+    homepage        https://github.com/edgars-supe/beets-importreplace
```
Comment (reneeotten) [inline audio/beets/Portfile:376]: the `homepage` is set already by the github PG

File: audio/beets/Portfile [hunk 11]
```diff
@@ -0,0 +1,601 @@
...
+    patch.pre_args  -p1
+    patchfiles      moveall-2to3.diff
+
+    build {}
+
+    destroot {
+        xinstall -d ${destroot}${prefix}/share/${name}
+        xinstall -m 0644 ${worksrcpath}/README.md ${destroot}${prefix}/share/${name}
+
+        xinstall -d ${destroot}${python.pkgd}/beetsplug
```
Comment (reneeotten) [inline audio/beets/Portfile:434]: isn't this directory already created by the `beets` main port, which is a dependency of this one?

File: audio/beets/Portfile [hunk 12]
```diff
@@ -0,0 +1,601 @@
...
+    name            beets-noimport
+    version         0.1.0b5
+    revision        0
+
+    license         MIT
+
+    description     Beets plugin to add directories to the incremental import 'do not import' list
+    long_description ${description}
+
+    homepage        http://www.github.com/ttsda/beets-noimport
```
Comment (reneeotten) [inline audio/beets/Portfile:449]: please use `https`

File: audio/beets/Portfile (hunk 4 above)
Comment (reneeotten) [reply audio/beets/Portfile:506]: can you please change it then to `MIT` ;) ?

File: audio/beets/Portfile [hunk 13]
```diff
@@ -0,0 +1,601 @@
...
+    name            beets-alternatives
+    version         0.10.2.post1
+    revision        0
+
+    license         MIT
+
+    description     beets plugin to manage multiple files
+    long_description ${description}
+
+    homepage        http://www.github.com/geigerzaehler/beets-alternatives
```
Comment (reneeotten) [inline audio/beets/Portfile:104]: please use `https` to avoid redirects

File: audio/beets/Portfile [hunk 14]
```diff
@@ -0,0 +1,601 @@
...
+    name            beets-check
+    version         0.13.0
+    revision        0
+
+    license         MIT
+
+    description     beets plugin verifying file integrity with checksums
+    long_description ${description}
+
+    homepage        http://www.github.com/geigerzaehler/beets-check
```
Comment (reneeotten) [inline audio/beets/Portfile:254]: please use `https`

**Review by reneeotten — CHANGES_REQUESTED** [review]
sorry @catap things got busy.... I apologize for the remaining (nitpick) requests, but after that I think it should be all good to go and I'll merge this so that others can benefit from the large amount of work you did here!

File: audio/beets/Portfile (hunk 9 above)
Comment (catap) [reply audio/beets/Portfile:140]: @reneeotten this is a wired plugin which hasn't use any way to install it. It is just `.py` file and `Readme`.

Without this line build is failed with:
```
Error: Failed to destroot beets-amazon: xinstall: Unable to create new file for: /opt/local/var/macports/build/_Users_catap_src_macports-ports_audio_beets/beets-amazon/work/destroot/opt/local/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/beetsplug, No such file o
```

File: audio/beets/Portfile (hunk 11 above)
Comment (catap) [reply audio/beets/Portfile:449]: Same here, without this folder => build crashed.

File: audio/beets/Portfile (hunk 9 above)
Comment (reneeotten) [reply audio/beets/Portfile:162]: Yeah, true... the directory is created by the `beets` port, but you do  need to make if first in the `destroot` here.

**Comment by catap** [comment]
@reneeotten no apologize! This is huge `Portfile` and it is a good idea to test it for a while.

During my tests I've discovered a few missed (=optional) dependencies which was mentored only inside docs and not inside `setup.py` ;)

Also, I figured out how to introduce `-devel` port which that is also good idea because `beets` is releasing not often. Usually a year or two between releases and scope between releases huge.

**Comment by catap** [comment]
@reneeotten I guess I need your help. How can I fix last errors which is produced by `--nitpick`? The error:
```
Error: Portfile directory beets does not match port name beets-barcode
Error: Portfile directory beets does not match port name beets-follow
Error: Portfile directory beets does not match port name beets-importreplace
Error: Portfile directory beets does not match port name beets-originquery
Error: Portfile directory beets does not match port name beets-usertag
```

All this subports are using `github.setup`.

File: audio/beets/Portfile [hunk 15]
```diff
@@ -0,0 +1,618 @@
...
+
+subport ${name}-full {
+    python.rootname beets
+    revision        0
+
+    description     Install beets with all plugins.
+    long_description \
+                    ${description}
+
+
```
Comment (reneeotten) [inline audio/beets/Portfile:584]: please add: 
```
    fetch           {}
    extract         {}
    checksum        {}

    livecheck.type  none
```
and then you can remove the `python.rootname` above

**Review by reneeotten — COMMENTED** [review]
@catap hhm... I am not sure how to resolve the issue about the Portfile directory and the `name`, which is set by the python PG in these subports. Ideally, you would not change the name from the "main" port - but in this case there isn't much one can do about it I think... So for now I'd suggest to just ignore it. 

I'd suggest a few more additions to the `beets-full` stub port to make sure it's not trying to any fetch/extract/checksum/livecheck as this doesn't make sense for a stub port. Other than that I think we would be ready to merge this - I would like to make sure that everything builds happily on the CI systems, but I am afraid it will timeout.

However, I assume that you have build everything locally and that this all works as intended. In any case, if it's merged and some issue comes up it can always be resolved at that point!

**Comment by catap** [comment]
@reneeotten yes, I'm using it locally which is installed as `beets-devel beets-full` and it is working fine.

And yes, I absolutley agree that better to merge it because pulish it can be done after and it will require sometime.

Regarding `name` issue. I feel that the right way is introducing `github.no_setup_name` and `python.no_setup_name` options and just enable it inside conflicted subports to allow `subport` take care of `name`. Another option will be split it to multiple port files that I'd like to avoid.

**Comment by catap** [comment]
@reneeotten yes, https://github.com/macports/macports-ports/pull/12289 is enough to pass `port lint --nitpick` :)
```
--->  Verifying Portfile for beets
--->  0 errors and 0 warnings found.
--->  Verifying Portfile for beets-alternatives
--->  0 errors and 0 warnings found.
--->  Verifying Portfile for beets-amazon
--->  0 errors and 0 warnings found.
--->  Verifying Portfile for beets-artistcountry
--->  0 errors and 0 warnings found.
--->  Verifying Portfile for beets-autofix
--->  0 errors and 0 warnings found.
--->  Verifying Portfile for beets-barcode
--->  0 errors and 0 warnings found.
--->  Verifying Portfile for beets-beetcamp
--->  0 errors and 0 warnings found.
--->  Verifying Portfile for beets-bpmanalyser
--->  0 errors and 0 warnings found.
--->  Verifying Portfile for beets-check
--->  0 errors and 0 warnings found.
--->  Verifying Portfile for beets-copyartifacts
--->  0 errors and 0 warnings found.
--->  Verifying Portfile for beets-describe
--->  0 errors and 0 warnings found.
--->  Verifying Portfile for beets-devel
--->  0 errors and 0 warnings found.
--->  Verifying Portfile for beets-follow
--->  0 errors and 0 warnings found.
--->  Verifying Portfile for beets-full
--->  0 errors and 0 warnings found.
--->  Verifying Portfile for beets-goingrunning
--->  0 errors and 0 warnings found.
--->  Verifying Portfile for beets-ibroadcast
--->  0 errors and 0 warnings found.
--->  Verifying Portfile for beets-importreplace
--->  0 errors and 0 warnings found.
--->  Verifying Portfile for beets-mosaic
--->  0 errors and 0 warnings found.
--->  Verifying Portfile for beets-moveall
--->  0 errors and 0 warnings found.
--->  Verifying Portfile for beets-noimport
--->  0 errors and 0 warnings found.
--->  Verifying Portfile for beets-originquery
--->  0 errors and 0 warnings found.
--->  Verifying Portfile for beets-summarize
--->  0 errors and 0 warnings found.
--->  Verifying Portfile for beets-usertag
--->  0 errors and 0 warnings found.
--->  Verifying Portfile for beets-xtractor
--->  0 errors and 0 warnings found.
--->  Verifying Portfile for beets-yearfixer
--->  0 errors and 0 warnings found.
```

**Comment by reneeotten** [comment]
> @reneeotten yes, #12289 is enough to pass `port lint --nitpick` with enabled hack at few subports :)

I am not convinced that that PR will get merged since this is quite a specific issue (and perhaps there is another way of resolving this, but I am not seeing it right now). In any case, let's just ignore that message about the `name` for now. I'll merge this - it's unfortunately not building for some reason on the CI but whatever I've tried locally does work. I am going to merge this now - thanks for all the work @catap!

**Comment by catap** [comment]
@reneeotten thanks!

Let me when migrate to #12289 example of changes to start discussion.

**Pull request merged**